### PR TITLE
Add utilities functions

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @gkoz @EPashkin 

@jo: It should fulfill your needs.

### CODE REVIEW

**Comment by jo** [comment]
@GuillaumeGomez awesome!

How can I test this? Would it be enough to point to this branch from Cargo.toml?

Thank you so much.

**Comment by GuillaumeGomez** [comment]
When merged, you can just put the glib dependency like this:

```toml
glib = { git = "https://github.com/gtk-rs/glib" }
```

To test it, you have to make it point to my own repo:

```toml
glib = { git = "https://github.com/guillaumegomez/glib", branch = "missing_functions" }
```

**Comment by jo** [comment]
I get an error:
```
[jo@jolap simple-gtk-application]$ cargo run
    Updating git repository `https://github.com/guillaumegomez/glib`
    Updating git repository `https://github.com/gtk-rs/sys`
error: native library `glib` is being linked to by more than one package, and can only be linked to by one package

  glib-sys v0.3.1
  glib-sys v0.3.1 (https://github.com/gtk-rs/sys#774929a2)
```

**Comment by jo** [comment]
My Cargo.toml looks like this:
```toml
[package]
name = "simple-gtk-application"
version = "0.1.0"
authors = ["Johannes J. Schmidt <jo@die-tf.de>"]

[dependencies]
glib = { git = "https://github.com/guillaumegomez/glib", branch = "missing_functions" }
gio = "0.1.0"
hyper = "0.9.10"
serde = "0.8"
serde_json = "0.8"
serde_derive = "0.8"

[dependencies.gtk]
version = "0.1.0"
features = ["v3_12"]
```

**Comment by jo** [comment]
Should I point the other gtk-rs deps to git#master?

**Comment by GuillaumeGomez** [comment]
Hum... I think it's because of gtk. Just fork our gtk and update its glib dependency like I showed you and then put your gtk dependency on your repo.

Sorry to have you put through this. :-/

**Comment by GuillaumeGomez** [comment]
You don't use gtk? Well, what I said for gtk is still valid for gio.

**Comment by jo** [comment]
No worry, I am so happy you help!

Yes, I do use gtk, look at the last section of my Cargo. Do I also need to for gio?

**Comment by jo** [comment]
Oh, the problems continue:
```
[jo@jolap gtk]$ cargo build
    Updating git repository `https://github.com/gtk-rs/sys`
    ...

error[E0432]: unresolved import `glib::signal::Inhibit`
 --> src/signal.rs:8:9
  |
8 | pub use glib::signal::Inhibit;
  |         ^^^^^^^^^^^^^^^^^^^^^ no `Inhibit` in `glib::signal`
```

**Comment by jo** [comment]
Those cross repository dependencies are always a pain. I know this from node so well :(

**Comment by GuillaumeGomez** [comment]
Sorry, my branch wasn't up-to-date with master. Refresh and retry.

**Comment by jo** [comment]
How do I refresh? I deleted target directory but with same errors

**Comment by jo** [comment]
This is my fork of gtk: https://github.com/jo/gtk
I did the following changes:
```sh
[jo@jolap gtk]$ git diff
```
```diff
diff --git a/Cargo.toml b/Cargo.toml
index f3d959c..1f39a68 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -71,8 +71,11 @@ git = "https://github.com/gtk-rs/gio"
 version = "0.1.0"
 
 [dependencies.glib]
-git = "https://github.com/gtk-rs/glib"
-version = "0.1.0"
+# git = "https://github.com/gtk-rs/glib"
+# version = "0.1.0"
+git = "https://github.com/guillaumegomez/glib"
+branch = "missing_functions"
+
 
 [dependencies.gdk]
 git = "https://github.com/gtk-rs/gdk"
```

**Comment by GuillaumeGomez** [comment]
It should be ok now, no?

**Comment by jo** [comment]
no :(
Maybe Cargo did not update the git checkout of your branch? How can I enforce this?

**Comment by GuillaumeGomez** [comment]
Well, it's not convenient to try to explain. So for now, please clone the repo locally and replace glib dependencies with `glib = { path = "../glib" }` (if glib and gtk are at the same level). Then from your project, replace gtk dependency with `gtk = { path = "../gtk" }` (if gtk is at the same level than your project).

I hope it'll make things easier while we wait this PR to get merged.

**Comment by jo** [comment]
OK, will try this during the next days.
Now I learned how to use local dependencies, thank you so much!

**Comment by GuillaumeGomez** [comment]
No problem. Like I said, once merged, you won't have to trouble yourself with all this.

File: src/enums.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        match *self {
+            UserDirectory::Desktop => ffi::G_USER_DIRECTORY_DESKTOP,
+            UserDirectory::Documents => ffi::G_USER_DIRECTORY_DOCUMENTS,
+            UserDirectory::Downloads => ffi::G_USER_DIRECTORY_DOWNLOAD,
+            UserDirectory::Music => ffi::G_USER_DIRECTORY_MUSIC,
+            UserDirectory::Pictures => ffi::G_USER_DIRECTORY_PICTURES,
+            UserDirectory::PublicShare => ffi::G_USER_DIRECTORY_PUBLIC_SHARE,
+            UserDirectory::Templates => ffi::G_USER_DIRECTORY_TEMPLATES,
+            UserDirectory::Videos => ffi::G_USER_DIRECTORY_VIDEOS,
+            UserDirectory::NDirectories => ffi::G_USER_N_DIRECTORIES,
```
Comment (EPashkin) [inline src/enums.rs]: We really need this case in manual code?

File: src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+pub fn path_skip_root(file_name: &str) -> Option<String> {
+    unsafe {
+        from_glib_none(ffi::g_path_skip_root(file_name.to_glib_none().0))
+    }
+}
+
+pub fn path_get_basename(file_name: &str) -> Option<String> {
+    unsafe {
+        from_glib_none(ffi::g_path_get_basename(file_name.to_glib_none().0))
```
Comment (EPashkin) [inline src/utils.rs]: `Returns: a newly allocated string containing the last component of the filename. ` => `from_glib_full`?

File: src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    }
+}
+
+pub fn path_get_dirname(file_name: &str) -> Option<String> {
+    unsafe {
+        from_glib_none(ffi::g_path_get_dirname(file_name.to_glib_none().0))
+    }
+}
+
+pub fn find_program_in_path(program: &str) -> Option<String> {
```
Comment (EPashkin) [inline src/utils.rs]: Also `from_glib_full`

File: src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+pub fn set_prgname(name: Option<&str>) {
+    unsafe {
+        ffi::g_set_prgname(name.to_glib_none().0)
+    }
+}
+
+pub fn get_environ() -> Vec<String> {
+    unsafe {
+        FromGlibPtrContainer::from_glib_none(ffi::g_get_environ())
```
Comment (EPashkin) [inline src/utils.rs]: Also `transfer full` but not sure which case.

File: src/enums.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/enums.rs]: Well, it's part of the enum so yes.

File: src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
+// Copyright 2015-2016, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use ffi;
+use translate::*;
+use UserDirectory;
+
+pub fn get_application_name() -> String {
```
Comment (GuillaumeGomez) [inline src/utils.rs]: Good catch!

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/utils.rs (hunk 5 above)
Comment (EPashkin) [inline src/utils.rs]: But it still `String`

**Comment by GuillaumeGomez** [comment]
Should be good now.

**Comment by EPashkin** [comment]
👍 
But I still not like `NDirectories` in rust code.

**Comment by GuillaumeGomez** [comment]
Me neither but it doesn't harm to have it so...

**Comment by jo** [comment]
@GuillaumeGomez I give up trying to compile gtk-rs packages with local dependencies. I ended up with clones of cairo, gdk, gdk-pixbuf, gio, glib, gtk, pango and sys, used path references everywhere but I get the error:
```
[jo@jolap gdk]$ cargo build
error: native library `cairo` is being linked to by more than one package, and can only be linked to by one package

  cairo-sys-rs v0.3.1 (https://github.com/gtk-rs/cairo#4a61ff90)
  cairo-sys-rs v0.3.1 (file:///home/jo/github/gtk-rs/cairo/cairo-sys-rs)
```
This error does not tell me where it happens :/

I will wait for this pr to get landed. Its not urgent for me anyway, I just wanted to contribute my testing. Although its interesting how the libs work together.

**Comment by GuillaumeGomez** [comment]
I'll merge it for now. If we encounter issues, we'll just update.

**Comment by jo** [comment]
\o/ it works: when I now set
```rs
glib::utils::set_program_name(Some("Simple Gtk App"));
```
it shows the name in the menubar:
bildschirmfoto vom 2016-11-06 11-52-40

Without `set_program_name` it just displays a dot:
bildschirmfoto vom 2016-11-06 11-56-54

Thank you so much for your work on gtk-rs!

**Comment by GuillaumeGomez** [comment]
Great!

**Comment by gkoz** [comment]
I'd recommend removing anything that's already available via `std::env`, `std::fs`, `std::net`. GLib makes sense as a standard library for C, but not for Rust.

**Pull request merged**